numba/cuda/cudadrv/driver.py

```diff
@@ -899,32 +893,39 @@ def load_module_image(context, image):
                   _module_finalizer(context, handle))
 
 
-def _make_mem_finalizer(dtor, bytesize):
-    def mem_finalize(context, handle):
-        allocations = context.allocations
-        deallocations = context.deallocations
+def _alloc_finalizer(context, handle, size):
+    allocations = context.allocations
+    deallocations = context.deallocations
 
-        def core():
-            if allocations:
-                del allocations[handle.value]
-
-            deallocations.add_item(dtor, handle, size=bytesize)
+    def core():
+        if allocations:
+            del allocations[handle.value]
+        deallocations.add_item(driver.cuMemFree, handle, size)
 
-        return core
+    return core
 
-    return mem_finalize
 
+def _hostalloc_finalizer(context, handle, size, mapped):
```

Comment:
Can you add a comment here about the delayed deallocation and how this finalizer relates to `_pin_finalizer`?